nitrates/analysis_seeds/UpperLimits.py

```diff
@@ -0,0 +1,308 @@
+    ra1, dec1 = convert_theta_phi2radec(100,100,att_q)
+    print('test ra, dec=',ra1,dec1)
+
+    ebins0 = np.array([15.0, 24.0, 35.0, 48.0, 64.0])
+    ebins0 = np.append(ebins0, np.logspace(np.log10(84.0), np.log10(500.0), 5+1))[:-1]
+    ebins0 = np.round(ebins0, decimals=1)[:-1]
+    ebins1 = np.append(ebins0[1:], [350.0])
+    nebins = len(ebins0)
+    logging.debug("Number of ebins: ")
+    logging.debug(nebins)
+
+    # number of detectors being used
+    ndets = np.sum(dmask==0)
+    logging.debug("Ndets: ")
+    logging.debug(np.sum(dmask==0))
+
+    t_end = trigger_time + 1e3
+    t_start = trigger_time - 1e3
+    mask_vals = mask_detxy(dmask, ev_data)
+
+    logging.debug(mask_vals)
+    bl_dmask = (dmask==0.)
+
+    # get rid of events:
+    # far away from trigger time
+    # from bad dets
+    # with bad event flags
+    bl_ev = (ev_data['EVENT_FLAGS']<1)&\
+        (ev_data['ENERGY']<=500.)&(ev_data['ENERGY']>=14.)&\
+        (mask_vals==0.)&(ev_data['TIME']<=t_end)&\
+        (ev_data['TIME']>=t_start)
+
+    #logging.info("Number of events passing cuts: ", np.sum(bl_ev))
+    ev_data0 = ev_data[bl_ev]
+
+   # logging.info(GTI_PNT)
+
+   # for row in GTI_PNT:
+   #     print(row['START'] - trigger_time, row['STOP'] - trigger_time)
+
+        
+    bkg_fname = os.path.join(work_dir,'bkg_estimation.csv')
+    bkg_df = pd.read_csv(bkg_fname)
+
+
+# create bkg object
+
+    try:
+        GTI = Table.read(evfname, hdu="GTI_POINTING")
+    except:
+        GTI = Table.read(evfname, hdu="GTI")
+    tmin = GTI_PNT["START"][0]
+    tmax = GTI_PNT["STOP"][-1]
+    twind = 20
+# print(GTI,tmin,tmax,twind)
+
+    poly_trng = np.int(twind)
+
+    bkg_obj = Linear_Rates(ev_data, tmin, tmax, trigger_time, GTI_PNT, sig_clip=4.0, poly_trng=poly_trng)
+#print(bkg_obj)
+    bkg_obj.do_fits()
+
+
+
+# Getting rate and std deviations for all durations:
+
+    dur_values=[0.128, 0.256, 0.512, 1.024, 2.048, 4.096, 8.192, 16.384]
+    #dur_values=[0.128] 
+    for k in range(len(dur_values)):
+
+# calculate bkg rate and rate error
+        twind = 20
+        dtmin = -20
+        dtmax = twind
+
+        tstep = dur_values[k] / 4.0
+        tbins0 = np.arange(dtmin, dtmax, tstep) + trigger_time
+        tbins1 = tbins0 + dur_values[k]
+        tcnts = get_cnts_tbins_fast(tbins0, tbins1, ev_data)
+
+        ntbins = len(tbins0)
+        snrs = np.zeros(ntbins)
+        sig2_bkg_values =[]
+        
+        for i in range(ntbins):
+            dur = tbins1[i] - tbins0[i]
+            tmid = (tbins1[i] + tbins0[i]) / 2.0
+            bkg_rate, bkg_rate_err = bkg_obj.get_rate(tmid)
+            sig2_bkg = (bkg_rate_err * dur) ** 2 + (bkg_rate * dur)
+            sig2_bkg_values.append(sig2_bkg)
+            snrs[i] = (tcnts[i] - bkg_rate * dur) / np.sqrt(sig2_bkg)
+   
+        max_sig2_bkg = np.max(sig2_bkg_values)
+        #print('max of sig2_bkg=',max_sig2_bkg)
+        #print('sig2_bkg=',sig2_bkg)
+        logging.debug('max sig2_bkg:')
```

Comment:
another useful debug statement would be the min or mean of sig2_bkg_values, to know how much taking the max is making a difference, also to make sure it's not a huge outlier